docs/AUTH_PROTOCOL.md

```diff
@@ -0,0 +1,393 @@
+    - Response: `response` event with arguments:
+        - `text`: The text entered by the user
+
+- `newPassword`: UI for choosing a new password
+    - The frontend prompts for the user to enter a new password twice, and only responds once the two instances of the new password match.
+    - The frontend should use libpwquality to enforce the system-wide password quality policy before responding with the new password
+    - This method is always primary
+    - Arguments:
+        - `reason`: The reason that the user is being prompted to pick a new password. Options:
+            - `unset`: The user doesn't have a password set yet, so we're setting one now
+            - `expired`: The user's previous password has expired
+            - `custom`: Just always display the `fallbackMessage`
+        - `fallbackMessage`: The message to display if the UI doesn't know how to handle the `reason`
+    - Response: `response` event with arguments:
+        - `newPassword`: The user's new password
+
+- `pin`: UI for entering a numeric PIN code
+    - This method is always primary
+    - Arguments:
+        - `length`: The number of digits in the PIN code
+        - `overridePrompt`: Overrides the default prompt (or, more likely, adds a prompt) to the PIN entry UI. Optional. Should only be used if absolutely necessary
+    - Response: `reponse` event with arguments:
+        - `pin`: The PIN code entered by the user
+
+- `otp`: UI for entering a one-time-password
+    - This method is always primary
+    - Arguments:
+        - `alphanumeric`: Allow alphabetic characters in the OTP. Defaults to false
+        - `length`: The number of digits/characters in the OTP
+        - `prefix`: Some OTPs start with a prefix (e.g. Google likes to use `G-######` as their format). Defaults to empty
+        - `suffix`: Similar to `prefix`, but for a suffix instead
+        - `overridePrompt`: Overrides the default prompt (or, more likely, adds a prompt) to the OTP entry UI. Optional. Should only be used if absolutely necessary
+    - Response: `response` event with arguments:
+        - `otp`: The OTP entered by the user. Neither the prefix nor the suffix are included.
+
+- `chooser`: UI for choosing from a list of options
+    - This method is always primary
+    - Arguments:
+        - `prompt`: A string to display to the user, explaining what they're choosing between. Optional.
+        - `options`: An array of objects that describes the options that the user can select between. Each object in the array has the following keys:
+            - `key`: The machine-usable string that identifies this option
+            - `name`: The display name for this option, to be presented to the user
+    - Response: `response` event with arguments:
+        - `selection`: The selected option's `key`
+
+- `eidp`: Web-based SSO login
+    - The frontend displays a QR code and/or an embedded web browser that loads a web page for single-sign-on
```

Comment:
The authenticator needs to know if the frontend supports an embedded web browser or not, because that determines which OAuth 2.0 flow it can use.

If the frontend supports an embedded web browser, the authenticator can use the Authorization Code Flow which has fewer steps (it doesn't require the user to enter a device code), else it can only use the Device Authorization Flow.

So we might want to use different authentication mechanisms for these two cases, or we adapt the  message with the "supportedMechanisms" to include that information.

Reply:
Yes, embedded web view only makes sense with the Authorization Code Flow and login code and QR code only make sense with the Device Authorization Flow.

Reply:
Also, it doesn't need a 'Done' button. The auth module should be able to tell when the authentication in the browser has completed, and can indicate that via the api (so that the web view can be closed).

Reply:
Honestly, I think the webview is all that would need to be displayed. You don't even need to show the User name, etc, since that will be displayed in the web view.

Reply:
Hmm, we could add a `eidpSupportsEmbeddedBrowser` in the `hello` message